crates/shared/src/price_estimation/native/coingecko.rs

```diff
@@ -158,19 +178,35 @@ impl CoinGecko {
         token: &Token,
         denominator_price_eth: Decimal,
         prices: &HashMap<Token, f64>,
+        infos: &HashMap<Token, TokenInfo>,
+        denominator: &Denominator,
     ) -> NativePriceEstimateResult {
+        let token_decimals = infos
+            .get(token)
+            .and_then(|i| i.decimals)
+            .with_context(|| format!("missing decimals: {token:?}"))?;
+
         let token_price_eth: Decimal = prices
             .get(token)
             .cloned()
             .ok_or(PriceEstimationError::NoLiquidity)?
             .try_into()
             .context("failed to convert price to decimal")?;
 
-        Ok(token_price_eth
+        // When the quoted token and the denominator have different number of decimals
+        // the computed price effectively needs to be shifted by the difference.
+        let adjustment =
+            Decimal::new(10, 0).powi(denominator.decimals as i64 - token_decimals as i64);
```

Comment:
it could be rare, but 10 ^ 255 would result in panic, we could use `checked_powi` just in case

Reply:
What's more, please can you do the casting like `i64::from()` instead of `as`. The reason is that if we one day change the underlying type (`u8`) for whatever reason, we guarantee it will always have a safe casting (because if it doesn't exist the `From<>` trait then it will not compile).